galley/pkg/runtime/resource/resource.go

```diff
@@ -0,0 +1,109 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package resource contains core abstract types for representing configuration resources.
+package resource
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+// Kind is the name of a resource type.
+// TODO(https://github.com/istio/istio/issues/6434): We should rename this to resource.MessageName to
+// align it with the proto MessageName.
+type Kind string
+
+// Version is the version identifier of a resource.
+type Version string
+
+// Key uniquely identifies a (mutable) config resource in the config space.
+type Key struct {
+	// Kind of the resource.
+	Kind Kind
+
+	// Fully qualified name of the resource.
+	FullName string
+}
+
+// VersionedKey uniquely identifies a snapshot of a config resource in the config space, at a given
+// time.
+type VersionedKey struct {
+	Key
+	Version Version
+}
+
+// Entry is the abstract representation of a versioned config resource in Istio.
+type Entry struct {
+	ID   VersionedKey
+	Item proto.Message
+}
+
+// Info is the type metadata for an Entry.
+type Info struct {
+	// The kind of resource that this info is about
+	Kind Kind
+
+	// The Type URL to use, when encoding as Any
+	TypeURL string
```

Comment:
Do we really need TypeURL if we have Kind? (assuming Kind is the MessageName). We could abstract the over-the-wire encoding (via any) in the MCP layer and use MessageName throughout the runtime package.

Reply:
It looks like `TypeURL` is needed because MCP server's supportedTypes expects the fully qualified name. I should fix that so we can unify on MessageName. Created https://github.com/istio/istio/issues/6437 to track follow-up.

Reply:
The TypeURL should be getting used at the final encoding point only, so rest of the runtime code should be on the Kind(MessageName) plan. But otherwise, we should be obtaining the TypeURL from the metadata package. It keeps metadata & mappings in a single place, rather than being implicit & sprinkled through the code.